src/mbgl/gl/context.cpp

```diff
@@ -327,11 +262,189 @@ void Context::setDirtyState() {
     vertexArrayObject.setDirty();
 }
 
+void Context::clear(optional<mbgl::Color> color,
+                    optional<float> depth,
+                    optional<int32_t> stencil) {
+    GLbitfield mask = 0;
+
+    if (color) {
+        mask |= GL_COLOR_BUFFER_BIT;
+        clearColor = *color;
+        colorMask = { true, true, true, true };
+    }
+
+    if (depth) {
+        mask |= GL_DEPTH_BUFFER_BIT;
+        clearDepth = *depth;
+        depthMask = true;
+    }
+
+    if (stencil) {
+        mask |= GL_STENCIL_BUFFER_BIT;
+        clearStencil = *stencil;
+        stencilMask = 0xFF;
+    }
+
+    MBGL_CHECK_ERROR(glClear(mask));
+}
+
+#if not MBGL_USE_GLES2
+DrawMode Context::operator()(const Points& points) {
+    pointSize = points.pointSize;
+    return DrawMode::Points;
+}
+#else
+DrawMode Context::operator()(const Points&) {
+    return DrawMode::Points;
+}
+#endif // MBGL_USE_GLES2
+
+DrawMode Context::operator()(const Lines& lines) {
+    lineWidth = lines.lineWidth;
+    return DrawMode::Lines;
+}
+
+DrawMode Context::operator()(const LineStrip& lineStrip) {
+    lineWidth = lineStrip.lineWidth;
+    return DrawMode::LineStrip;
+}
+
+DrawMode Context::operator()(const Triangles&) {
+    return DrawMode::Triangles;
+}
+
+DrawMode Context::operator()(const TriangleStrip&) {
+    return DrawMode::TriangleStrip;
+}
+
+std::size_t Context::VertexArrayObjectHash::operator()(const VertexArrayObjectKey& key) const {
+    std::size_t seed = 0;
+    boost::hash_combine(seed, std::get<0>(key));
+    boost::hash_combine(seed, std::get<1>(key));
+    boost::hash_combine(seed, std::get<2>(key));
+    boost::hash_combine(seed, std::get<3>(key));
+    return seed;
+}
+
+void Context::setDepth(const Depth& depth) {
+    if (depth.func == Depth::Always && !depth.mask) {
+        depthTest = false;
+    } else {
+        depthTest = true;
+        depthFunc = depth.func;
+        depthMask = depth.mask;
+        depthRange = depth.range;
+    }
+}
+
+void Context::setStencil(const Stencil& stencil) {
+    if (stencil.test.is<Stencil::Always>() && !stencil.mask) {
+        stencilTest = false;
+    } else {
+        stencilTest = true;
+        stencilMask = stencil.mask;
+        stencilOp = { stencil.fail, stencil.depthFail, stencil.pass };
+        apply_visitor([&] (const auto& test) {
+            stencilFunc = { test.func, stencil.ref, test.mask };
+        }, stencil.test);
+    }
+}
+
+void Context::setColor(const Color& color) {
+    if (color.blendFunction.is<Color::Replace>()) {
+        blend = false;
+    } else {
+        blend = true;
+        blendColor = color.blendColor;
+        apply_visitor([&] (const auto& blendFunction) {
+            // TODO: blendEquation = blendFunction.equation;
```

Comment:
Previous code didn't call `glBlendEquation` anywhere and we always want the default of `GL_FUNC_ADD`. But we should set it explicitly right? In case a custom layer changes it.